### DISCUSSION TITLE
[DISCUSSTION]Alamofire를 이용한 네트워크 레이어 레퍼런스

### DISCUSSION THREAD

**Discussion by kimscastle** [Team-Plu/Plu-iOS#59] [General]
기존 프로젝트에서는 async/await과 urlsession을 통한 네트워크 레이어를 구축했습니다
### 현재 적용 예정인 네트워크 레이어 구조
- viewmodel -> manager -> service -> APIService
### 기존 프로젝트

- manager의 역할 : Response를 AppData로 변환
- service의 역할 : Response를 전달
- APIService의 역할 : API통신후 Data를 Response로 변환

### 변경될 레이어의 역할
- manager의 역할 : Response를 AppData로 변환
- service의 역할 : Data를 Response로 변환 후 반환 (decode)
- APIService의 역할 : API통신후 Data 반환

하지만 이번 프로젝트에서는 Alamofire과 async/await 그리고 combine을 활용한 네트워크 레이어를 구축하기로 결정을해서 많은 레퍼런스를 참고해서 하나의 레퍼런스를  결정해서 해당 방식으로 구현을 하는건 어떨까요

주말을 이용해서 여러 레퍼런스를 참고해서 본인이 괜찮다싶은 레퍼런스 링크와 특징을 기록해보시죠
> 필수적으로 들어가야하는 내용 = 레퍼런스링크, 본인이 생각한 특징

**Comment by kimscastle** [comment]
## EventMonitor
### 네트워킹통신결과를 일관성있게 출력하는 방법
<img width="792" alt="스크린샷 2024-01-05 오후 3 04 14" src="https://github.com/Team-Plu/Plu-iOS/assets/99013115/ca4b25b3-50a2-43ca-a517-f5eb4b3c287c">

- 프로토콜이라 해당 프로토콜을 채택해서 `requestDidFinish(_:)`와 `request<Value>(_:didParseResponse:)`얘네를 구현해주면됩니다
1. `requestDidFinish(_:)`은 request가 끝나고 response가 시작될때 호출
2. `request<Value>(_:didParseResponse:)`응답을 받은 경우 호출

이렇게하면 alamofire의 AF를 한번더 wrapping해줘야하는데 그렇게 하는김에 PluAPI이런식으로 묶으면 될거같네요

**Comment by ffalswo2** [comment]
## URLRequestConvertible
```swift
open func request(_ urlRequest: URLRequestConvertible, 
                  interceptor: RequestInterceptor? = nil) -> DataRequest
```
```swift
/// Types adopting the `URLRequestConvertible` protocol can be used to safely construct `URLRequest`s.
public protocol URLRequestConvertible {
    /// Returns a `URLRequest` or throws if an `Error` was encountered.
    ///
    /// - Returns: A `URLRequest`.
    /// - Throws:  Any error thrown while constructing the `URLRequest`.
    func asURLRequest() throws -> URLRequest
}

extension URLRequestConvertible {
    /// The `URLRequest` returned by discarding any `Error` encountered.
    public var urlRequest: URLRequest? { try? asURLRequest() }
}

extension URLRequest: URLRequestConvertible {
    /// Returns `self`.
    public func asURLRequest() throws -> URLRequest { self }
}
```
Alamofire의 request 파이프라인을 통과하는 모든 request의 기반으로 URLRequestConvertible을 사용하게 됩니다.
request과정에서 URLRequestConvertible의 asURLRequest가 호출되어 URLRequest를 생성합니다.

Alamofire의 AdvancedUsage에서는 해당 프로토콜과 Router Pattern을 같이 적절하게 활용하면 복잡한 앱의 API를 쉽게 관리하고 설계할 수 있다며 이를 권장하고 있습니다.
```swift
enum Router: URLRequestConvertible {
    case get([String: String]), post([String: String])

    var baseURL: URL {
        return URL(string: "https://httpbin.org")!
    }

    var method: HTTPMethod {
        switch self {
        case .get: return .get
        case .post: return .post
        }
    }

    var path: String {
        switch self {
        case .get: return "get"
        case .post: return "post"
        }
    }

    func asURLRequest() throws -> URLRequest {
        let url = baseURL.appendingPathComponent(path)
        var request = URLRequest(url: url)
        request.method = method

        switch self {
        case let .get(parameters):
            request = try URLEncodedFormParameterEncoder().encode(parameters, into: request)
        case let .post(parameters):
            request = try JSONParameterEncoder().encode(parameters, into: request)
        }

        return request
    }
}
```
URLRequestConvertible과 Router Pattern을 이용하면 우리 앱에서도 엔드포인트를 쉽게 관리할 수 있는 장점이 있어보입니다.

**Comment by kimscastle** [comment]
## 토큰갱신하는방법
https://ios-development.tistory.com/730?category=899471
- 로그인 auth쪽에서 토큰을 refresh하는방식을 적용해보면 좋을거같네용

**Comment by ffalswo2** [comment]
Alamofire를 사용할 때 Alamofire의 Request 파이프라인을 이해하고 있으면 여러므로 도움이 될 것 같아서 아래 링크 또 따로 올리겠습니다.

https://github.com/Alamofire/Alamofire/blob/master/Documentation/AdvancedUsage.md#requests

1. Initial parameters, like HTTP method, headers, and parameters are encapsulated into an internal URLRequestConvertible value. If a URLRequestConvertible value is passed directly, that value is used unchanged.
2. asURLRequest() is called on the the URLRequestConvertible value, creating the first URLRequest value. This value is passed to the Request and stored in requests. If the URLRequestConvertible value was created from the parameters passed to a Session method, any provided RequestModifier is called when the URLRequest is created.
3. If there are any Session or Request RequestAdapters or RequestInterceptors, they’re called using the previously created URLRequest. The adapted URLRequest is then passed to the Request and stored in requests as well.
4. Session calls the Request to create the URLSessionTask to perform the network request based on the URLRequest.
5. Once the URLSessionTask is complete and URLSessionTaskMetrics have been gathered, the Request executes its Validators.
6. Request executes any response handlers, such as responseDecodable, that have been appended.
